### PR TITLE
Open boundaries (again)

### PR DESCRIPTION
**Pull request by DanielRichtmann**
New PR since quite a lot has changed in the mean time, will close the old one in favor of this.

Features/checks:

- open bc for reference wilson clover operator
- open bc for c++ wilson clover operator
- explicit 1:1 check of most of the Dirac operator methods against openqcd designed to run in CI (apply methods in openqcd, write fields to disk, read fields with gpt, apply our methods, compare. uses [this](https://github.com/DanielRichtmann/openqcd/blob/f3304e46fd6582d44e8ae0a6b35d595e74e05dcf/devel/dirac/compare_gpt.c) openqcd binary, fields are in this [PR](https://github.com/lehner/gpt-repository/pull/1) to gpt-repository)

Offline checks of correlators against rqcd chroma with different solvers (cg, mp-fgmres-bicgstab, mg) were successful.
.
Questions/to discuss: 

- `Mdag` et. al. for reference operator?
- performance of `where` vs multiplication with boundary mask in c++ operator?
- more checks to run in CI, e.g., some form of the correlator check and/or integration in `fermion_operators.py`?

### CODE REVIEW

**Comment by lehner** [comment]
Thanks!  I am starting to go through this and have one request for the tests.  The test against openQCD should not become part of the CI (but should be available as a test that is not included in run).  Once verified against, e.g., openQCD we will rely on the fingerprint tests in https://github.com/lehner/gpt/blob/master/tests/qcd/fermion_operators.py to make sure that all matrix applications don't change by some screw up in the future.

**Comment by lehner** [comment]
We could meet tomorrow morning to discuss the open points.  Let's talk by email.

**Comment by DanielRichtmann** [comment]
> Thanks! I am starting to go through this and have one request for the tests. The test against openQCD should not become part of the CI (but should be available as a test that is not included in run). Once verified against, e.g., openQCD we will rely on the fingerprint tests in https://github.com/lehner/gpt/blob/master/tests/qcd/fermion_operators.py to make sure that all matrix applications don't change by some screw up in the future.

Sounds reasonable, I will push a commit removing this test from run but I would like to see CI run once before that. Even if we remove it from run, this test still requires the PR to gpt-repository to be merged.
I guess the fingerprint tests are what I was talking about in the last bullet.

File: lib/gpt/core/util.py [hunk 1]
```diff
@@ -87,3 +88,14 @@ def all_have_attribute(value, a):
     if type(value) == list and len(value) > 0:
         return all([all_have_attribute(v, a) for v in value])
     return hasattr(value, a)
+
+
+# fdimensions
+def fdimensions_from_openqcd(fdim):
```
Comment (lehner) [inline lib/gpt/core/util.py:94]: fdimensions_to/from_openqcd should not go into util.py .  We should never need them apart from a single test file.  They need to go into the test.

File: lib/gpt/core/covariant.py [hunk 2]
```diff
@@ -84,3 +84,19 @@ def __init__(self, U, params):
                 gpt.pick_checkerboard(cb, _U[mu], self.U[mu])
 
             self.checkerboard[cb] = shift_base(_U, None)
+
+
+def apply_open_boundaries(field):
```
Comment (lehner) [inline lib/gpt/core/covariant.py:89]: This function should not go into core/covariant.py .  Core does not know about QCD or any related jargon.  Why is open boundaries only a concept for >= dimensions and then only goes in the time direction?  The use in the multi-grid is also not elegant and we need to find a better general ansatz.

File: lib/gpt/algorithms/inverter/multi_grid.py [hunk 3]
```diff
@@ -136,6 +136,15 @@ def __init__(self, mat_f, params):
                 ]
             self.distribution[lvl](self.basis[lvl][0 : self.nb[lvl]])
 
+        # open bc for basis vectors if finest level matrix is open bc
+        # the way I retrieve this info is rather ugly -> TODO?
+        if self.mat[self.finest].params["boundary_phases"][-1] == 0.0:
```
Comment (lehner) [inline lib/gpt/algorithms/inverter/multi_grid.py:141]: Agree, we need a better way to retrieve this info but also open boundaries should not be hard-coded in a solver.  We need to understand the abstract idea and implement it.  The idea is likely just that the basis needs to be created differently, so it needs to be located with the basis creation?

File: lib/gpt/core/util.py (hunk 1 above)
Comment (DanielRichtmann) [reply lib/gpt/core/util.py]: Resolved in https://github.com/lehner/gpt/pull/79/commits/0242ad51a0d92431f27be78634de6541dbd727b7.

File: lib/gpt/core/covariant.py (hunk 2 above)
Comment (DanielRichtmann) [reply lib/gpt/core/covariant.py:89]: Outside of the Dirac operator, this file is the only place where the boundary phase (0 for open boundaries) gets multiplied into the gauge field. That's why I deemed it a good idea to put the analogue for the fermions in the same file. However, I don't have strong feelings about this. What would be a better place?
To be honest I don't understand the third sentence. Open boundaries as proposed by Lüscher only act in the temporal direction by construction.

File: lib/gpt/algorithms/inverter/multi_grid.py (hunk 3 above)
Comment (DanielRichtmann) [reply lib/gpt/algorithms/inverter/multi_grid.py:141]: This is the place where the basis vectors are filled with random numbers initially. On top of that open bc requires fermion fields to be zero at the temporal boundaries, which is what I do here. It's really that simple and I don't think there's a more abstract idea behind this.
The basis creation is located in the same file a few lines below this. However, moving this there would not improve on your points of criticism in my opinion.

Comment (lehner) [reply lib/gpt/algorithms/inverter/multi_grid.py:141]: 1) The solver should not know about parameters of the matrix it needs to invert.  Even assuming the existence of the params is poor design.

2) The solver *should* however consider itself with its own parameters, one of which is that the MG needs a prescription to go from coarse to fine and back or equivalently it needs a blocked basis.  The boundary phases are just a very special case of the abstract idea of how to create the blocked basis.

You can leave this to me, I will think about a solution and implement it.

File: lib/gpt/core/covariant.py (hunk 2 above)
Comment (lehner) [reply lib/gpt/core/covariant.py:89]: Open boundaries are a concept far beyond LQCD.  In the gpt/core module, we should not do things specific to LQCD and its jargon.  This is what the qcd module is for.  You can leave this also up to me.

**Pull request merged**